### PR TITLE
C++/Python API Parity: add AlphaDropout and FeatureAlphaDropout

### PR DESCRIPTION
**Pull request by suyashb95**
- add `AlphaDropoutImpl` to `modules/dropout.h` and `modules/dropout.cpp`
 - add `functional/dropout.h` containing the `alpha_dropout` function
 - include `functional/dropout.h` in `nn/functional.h`
 - add functional and module tests
-  related issue https://github.com/pytorch/pytorch/issues/25883

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/dropout.h [hunk 1]
```diff
@@ -64,6 +64,19 @@ class TORCH_API FeatureDropoutImpl
   void pretty_print(std::ostream& stream) const override;
 };
 
+class TORCH_API AlphaDropoutImpl
+    : public detail::DropoutImplBase<AlphaDropoutImpl> {
+ public:
+  explicit AlphaDropoutImpl(const DropoutOptions& options_ = DropoutOptions());
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/dropout.h]: @Suyash458 Thanks so much for the awesome work! I think we can change the arguments in `DropoutOptions` to include `p` and `inplace` instead of `rate` to match the Python version better. And we can have `using AlphaDropoutOptions = DropoutOptions;` and then use `AlphaDropoutOptions` everywhere in the AlphaDropout code :D

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include <torch/nn/options/dropout.h>
+
+namespace torch {
+namespace nn {
+namespace functional {
+
+inline Tensor alpha_dropout(const Tensor& input, const DropoutOptions& options, bool is_training) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: I believe we can write
```suggestion
inline Tensor alpha_dropout(const Tensor& input, const AlphaDropoutOptions& options = {}, bool training = false) {
```
to match the Python names better and also more consistent with the naming of other functional options :D

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  return torch::alpha_dropout(input, options.rate(), is_training);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: It would be awesome to express the same logic as in Python version:
```python
    if p < 0. or p > 1.:
        raise ValueError("dropout probability has to be between 0 and 1, "
                         "but got {}".format(p))
    return (_VF.alpha_dropout_(input, p, training)
            if inplace
            else _VF.alpha_dropout(input, p, training))
```

File: torch/csrc/api/src/nn/modules/dropout.cpp [hunk 4]
```diff
@@ -45,5 +46,16 @@ Tensor FeatureDropoutImpl::forward(const Tensor& input) {
 void FeatureDropoutImpl::pretty_print(std::ostream& stream) const {
   stream << "torch::nn::FeatureDropout(rate=" << options.rate() << ")";
 }
+
+AlphaDropoutImpl::AlphaDropoutImpl(const DropoutOptions& options_)
+    : DropoutImplBase(options_) {}
+
+Tensor AlphaDropoutImpl::forward(const Tensor& input) {
+  return torch::alpha_dropout(input, options.rate(), this->is_training());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/dropout.cpp]: I believe this can be
```suggestion
  return F::alpha_dropout(input, options.p(), this->is_training());
```
to match the Python version even better :D

File: torch/csrc/api/src/nn/modules/dropout.cpp [hunk 5]
```diff
@@ -45,5 +46,16 @@ Tensor FeatureDropoutImpl::forward(const Tensor& input) {
...
+}
+
+void AlphaDropoutImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::AlphaDropout(rate=" << options.rate() << ")";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/dropout.cpp]: We might want to print both `p` and `inplace`, to match the Python version even better :D

File: test/cpp/api/functional.cpp [hunk 6]
```diff
@@ -1267,3 +1267,15 @@ TEST_F(FunctionalTest, Pad) {
...
+TEST_F(FunctionalTest, AlphaDropout) {
+  auto input = torch::randn(5000);
+  auto input_mean = input.mean();
+  auto input_std = input.std();
+
+  for (const auto rate : {0.2, 0.5, 0.8}) {
+    auto output = F::alpha_dropout(input, rate, false);
+    ASSERT_TRUE(torch::allclose(input_mean, output.mean(), 0.1));
+    ASSERT_TRUE(torch::allclose(input_std, output.std(), 0.1));
+  }
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: Thanks so much for the awesome tests! It would be great if we could also add a test for the `F::alpha_dropout(input)` (i.e. no value passed for either `options` or `training`) use case :D

File: test/cpp/api/modules.cpp [hunk 7]
```diff
@@ -2079,3 +2079,11 @@ TEST_F(ModulesTest, PrettyPrintThreshold) {
...
     "torch::nn::Threshold(threshold=42.42, value=24.24, inplace=true)");
 }
+
+TEST_F(ModulesTest, PrettyPrintAlphaDropout) {
+  ASSERT_EQ(c10::str(AlphaDropout(0.2)),
+    "torch::nn::AlphaDropout(rate=0.2)");
+  ASSERT_EQ(c10::str(AlphaDropout(
+      DropoutOptions(0.2))),
+    "torch::nn::AlphaDropout(rate=0.2)");
+}
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: It would be awesome to add a module test similar to
https://github.com/pytorch/pytorch/blob/e31adeb4f38b41de5c0f840b6cdf05593711d4da/test/cpp/api/modules.cpp#L897-L911

File: torch/csrc/api/include/torch/nn/functional.h [hunk 8]
```diff
@@ -9,3 +9,4 @@
 #include <torch/nn/functional/pixelshuffle.h>
 #include <torch/nn/functional/pooling.h>
 #include <torch/nn/functional/vision.h>
+#include <torch/nn/functional/dropout.h>
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional.h]: nit: would be awesome to put this line above `#include <torch/nn/functional/embedding.h>`, following alphabetical order

**Review by yf225 — COMMENTED** [review]
@Suyash458 Thanks so much for the awesome work! I left some comments.

File: torch/csrc/api/include/torch/nn/modules/dropout.h (hunk 1 above)
Comment (Suyash458) [reply torch/csrc/api/include/torch/nn/modules/dropout.h]: @yf225 Thanks for having a look! Adding `p` and `inplace` makes sense, I'll add the extra params to `DropoutOptions`. This'll affect `DropoutImpl` and `FeatureDropoutImpl` as well

File: torch/csrc/api/include/torch/nn/functional/dropout.h (hunk 3 above)
Comment (Suyash458) [reply torch/csrc/api/include/torch/nn/functional/dropout.h]: Just wondering, wouldn't this bit of code take care of the check for the `p` value?

```c++
DropoutImplBase<Derived>::DropoutImplBase(const DropoutOptions& options_)
    : options(options_) {
  TORCH_CHECK(options.rate() >= 0, "Dropout rate must not be less than zero");
  TORCH_CHECK(options.rate() <= 1, "Dropout rate must not be greater than one");
}```

File: test/cpp/api/functional.cpp (hunk 6 above)
Comment (Suyash458) [reply test/cpp/api/functional.cpp]: Ah right, I missed that. Will add a test case for it

File: torch/csrc/api/include/torch/nn/modules/dropout.h (hunk 1 above)
Comment (yf225) [reply torch/csrc/api/include/torch/nn/modules/dropout.h]: Yes and I think we should rename `rate` to `p` in `DropoutOptions`, and it is okay to break backward compatibility in favor of code simplicity and parity with Python API

File: torch/csrc/api/include/torch/nn/functional/dropout.h (hunk 3 above)
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/dropout.h]: Yes I think it would take care of the check when the user is calling `F::alpha_dropout` from the `AlphaDropout` module, but the user might also call `F::alpha_dropout` directly, and they would be out of luck :D

File: test/cpp/api/modules.cpp (hunk 7 above)
Comment (Suyash458) [reply test/cpp/api/modules.cpp]: I have added a test similar to the one for `Dropout`. From what I read in [this](https://arxiv.org/pdf/1706.02515.pdf) paper, the sum of the output should lie between 40 and 130

>New Dropout Technique. Standard dropout randomly sets an activation x to zero with probability 1 − q for 0 < q 6 1 .  In order to preserve the mean, the activations are scaled by 1 /q during training.  If x has mean E(x) = μ and variance Var(x) = ν , and the dropout variable d follows a binomial distribution B(1 ,q ) , then the mean E(1 /qdx ) = μ is kept.  Dropout fits well to rectified linear units, since zero is in the low variance region and corresponds to the default value. For scaled exponential linear units, the default and low variance value is lim x →−∞ selu(x)  = − λα = α ′ . Therefore, we propose “alpha dropout”, that randomly sets inputs to α ′ .  The new mean and new variance is E(xd + α ′(1 − d)) = qμ + (1 − q)α ′ , and Var(xd + α ′(1 − d)) = q((1 − q)(α ′ − μ) 2 + ν) .  We aim at keeping mean and variance to their original values after “alpha dropout”, in order to ensure the self-normalizing property even for “alpha dropout”. The affine transformation a (xd + α ′ (1 − d)) + b allows to determine parameters a and b such that mean and variance are kept to their values: E( a ( xd + α ′ (1 − d )) + b ) = μ and Var( a ( xd + α ′ (1 − d )) + b ) = ν . In contrast to dropout, a and b will depend on μ and ν , however our SNNs converge to activations with zero mean and unit variance.  With μ = 0 and ν = 1 , we obtain a = (q + α′2q(1 − q )) − 1 / 2 and b = − ( q + α ′ 2 q (1 − q ) ) − 1 / 2 ((1 − q ) α ′ ) . The parameters a and b only depend on the dropout rate 1 − q and the most negative activation α

File: torch/csrc/api/include/torch/nn/functional/dropout.h (hunk 3 above)
Comment (Suyash458) [reply torch/csrc/api/include/torch/nn/functional/dropout.h]: Right, I've added the same checks in the `F::alpha_dropout` function

**Comment by Suyash458** [comment]
@yf225 I have added another commit with the required changes, lemme know if they look good!

File: test/cpp/api/functional.cpp (hunk 6 above)
Comment (yf225) [reply test/cpp/api/functional.cpp]: Thanks a lot for your help!

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace torch {
+namespace nn {
+namespace functional {
+
+using AlphaDropoutOptions = DropoutOptions;
+
+inline Tensor alpha_dropout(const Tensor& input, const AlphaDropoutOptions& options = {}, bool training = false) {
+  TORCH_CHECK(options.p() >= 0, "Dropout rate must not be less than zero");
+  TORCH_CHECK(options.p() <= 1, "Dropout rate must not be greater than one");
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: To strictly match the Python version, maybe we could do the following instead:
```cpp
TORCH_CHECK(
  options.p() >= 0 && options.p() <= 1,
  "dropout probability has to be between 0 and 1, but got ", options.p());
```

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  
+  return torch::alpha_dropout(input, options.p(), training);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: In the Python version, I think we execute different code paths based on the value of `options.inplace()`:
```python
    return (_VF.alpha_dropout_(input, p, training)
            if inplace
            else _VF.alpha_dropout(input, p, training))
```
It would be awesome to do the same in C++ version as well (we might need to change the type of `input` from `const Tensor&` to `Tensor` in order to achieve this) :D

File: torch/csrc/api/include/torch/nn/options/dropout.h [hunk 11]
```diff
@@ -4,16 +4,18 @@
 #include <torch/csrc/WindowsTorchApiMacro.h>
 #include <torch/types.h>
 
+
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/dropout.h]: nit: we can remove the new line here

File: torch/csrc/api/include/torch/nn/options/dropout.h [hunk 12]
```diff
@@ -4,16 +4,18 @@
...
 namespace torch {
 namespace nn {
 
 /// Options for `Dropout` and `FeatureDropout`.
 struct TORCH_API DropoutOptions {
-  /* implicit */ DropoutOptions(double rate = 0.5);
+  /* implicit */ DropoutOptions(double p = 0.5, bool inplace = false);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/dropout.h]: I think we can put the default value for `inplace` as part of the `TORCH_ARG` declaration instead, something like
```cpp
TORCH_ARG(bool, inplace) = false;
```
instead of specifying it as default value to the constructor argument. (Ideally we should do the same for `p` as well, but there are probably too many people using `Dropout(0.5)` directly, and we have to preserve it for better backward compatibility.)

File: torch/csrc/api/src/nn/modules/dropout.cpp [hunk 13]
```diff
@@ -8,42 +8,62 @@
...
+using AlphaDropoutOptions = DropoutOptions;
+
 namespace detail {
 template <typename Derived>
 DropoutImplBase<Derived>::DropoutImplBase(const DropoutOptions& options_)
     : options(options_) {
-  TORCH_CHECK(options.rate() >= 0, "Dropout rate must not be less than zero");
-  TORCH_CHECK(options.rate() <= 1, "Dropout rate must not be greater than one");
+  TORCH_CHECK(options.p() >= 0, "Dropout rate must not be less than zero");
+  TORCH_CHECK(options.p() <= 1, "Dropout rate must not be greater than one");
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/dropout.cpp]: To strictly match the Python version, maybe we could do the following instead:
```cpp
TORCH_CHECK(
  options.p() >= 0 && options.p() <= 1,
  "dropout probability has to be between 0 and 1, but got ", options.p());
```

File: test/cpp/api/modules.cpp (hunk 7 above)
Comment (yf225) [reply test/cpp/api/modules.cpp]: Thanks so much for adding the test! :D

**Review by yf225 — COMMENTED** [review]
@Suyash458 Thanks so much for the update! I left some minor comments.

**Comment by yf225** [comment]
We will probably need to rebase the PR because of the conflicts :(

File: torch/csrc/api/include/torch/nn/functional/dropout.h (hunk 10 above)
Comment (Suyash458) [reply torch/csrc/api/include/torch/nn/functional/dropout.h]: In `Functions.h` I see two declarations for `alpha_dropout`
``` cpp
static inline Tensor & alpha_dropout_(Tensor & self, double p, bool train) {...}
static inline Tensor alpha_dropout(const Tensor & self, double p, bool train) {...}
```
Guessing we have to use the first one for in-place operation?

**Comment by Suyash458** [comment]
Rebasing to resolve conflicts in test files

**Comment by Suyash458** [comment]
@yf225 is the current build failing on Linux? I get an error saying 
```
../caffe2/opt/onnxifi_op.cc: In function ‘void caffe2::{anonymous}::copyDescriptor(const caffe2::ExternalTensorDescriptor*, onnxTensorDescriptorV1*)’:
../caffe2/opt/onnxifi_op.cc:95:7: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
   to->isOffline = from->isOffline;
```
[Here's](https://pastebin.com/c75LN3CV) the complete stack trace

**Comment by Suyash458** [comment]
@yf225 lemme know if this looks fine, I'll rebase to fix the conflicts then

**Review by yf225 — APPROVED** [review]
@Suyash458 Thanks a lot for the awesome work!

**Comment by Suyash458** [comment]
> @Suyash458 Thanks a lot for the awesome work!

Glad I could help!
